### DISCUSSION TITLE
Parent extended public key is the same as child extended public key?

### DISCUSSION THREAD

**Discussion by mark2021blue** [trustwallet/wallet-core#2081] [Q&A]
**Describe the bug**
Parent extended public key is the same as child extended public key?

**Expected behavior**
Are the two xpubs "m/44'/3'/0'/0/0" and "m/44'/3'/1'/0/0" the same correct code?
My understanding:
"m/44'/3'/0'/0/0" TWHDWalletGetExtendedPublicKey Gets the parent extended public key
"m/44'/3'/1'/0/0" TWAccountExtendedPublicKey Get child extended public key

**Additional context**
auto secretMnemonict1 = TWStringCreateWithUTF8Bytes(mnemonic1);
            TWHDWallet* walletImpt1 = TWHDWalletCreateWithMnemonic(secretMnemonict1, TWStringCreateWithUTF8Bytes(""));
            TWStringDelete(secretMnemonict1);

            auto customDerivationPatht1 = TWStringCreateWithUTF8Bytes("m/44'/3'/0'/0/0");
            TWPrivateKey* secretPrivateKeyCustomt1 = TWHDWalletGetKey(walletImpt1, TWCoinType::TWCoinTypeDogecoin, customDerivationPatht1);
            TWStringDelete(customDerivationPatht1);
            string xpub = TWStringUTF8Bytes(TWHDWalletGetExtendedPublicKey(walletImpt1, TWPurposeBIP44, TWCoinTypeDogecoin, TWHDVersionDGUB));
            cout << "+++++++ xpub1:" << xpub << endl;

            auto customDerivationPath2 = TWStringCreateWithUTF8Bytes("m/44'/3'/1'/0/0");
            TWPrivateKey* secretPrivateKeyCustom2 = TWHDWalletGetKey(walletImpt1, TWCoinType::TWCoinTypeDogecoin, customDerivationPath2);
          //  TWStringDelete(customDerivationPath2);
            string addressCustom2 = TWStringUTF8Bytes(TWCoinTypeDeriveAddress(TWCoinType::TWCoinTypeDogecoin, secretPrivateKeyCustom2));
            cout << "+++++++++++ doge Custom-derived address:   '" << addressCustom2 << "'" << endl;

            TWString* aC2 = TWCoinTypeDeriveAddress(TWCoinType::TWCoinTypeDogecoin, secretPrivateKeyCustom2);
            TWString* pxpub = TWHDWalletGetExtendedPublicKey(walletImpt1, TWPurposeBIP44, TWCoinTypeDogecoin, TWHDVersionDGUB);
            TWAccount* accountT1 = TWAccountCreate(aC2, TWCoinType::TWCoinTypeDogecoin, customDerivationPath2, pxpub);
          //  TWAccountExtendedPublicKey(accountT1);

            string xpub2 = TWStringUTF8Bytes(TWAccountExtendedPublicKey(accountT1));
            string adp2 = TWStringUTF8Bytes(TWAccountDerivationPath(accountT1));
            cout << "+++++++ xpub2:" << xpub2 << " Custom-derived address:   '" << TWStringUTF8Bytes(aC2) << "' adp:" << adp2 << endl;
            cout << endl;

**Comment by optout21** [comment]
In the line 
```
TWString* pxpub = TWHDWalletGetExtendedPublicKey(walletImpt1, TWPurposeBIP44, TWCoinTypeDogecoin, TWHDVersionDGUB);
```
`TWHDWalletGetExtendedPublicKey` uses the default derivation path, that is `m/44'/3'/0'/0/0` (and not `m/44'/3'/1'/0/0`).
Therefore in the next line in `TWAccountCreate` the provided derivation path and public key are not consistent.
If you compare derivation of `xpub` and `pxpub`, they are the same line, so they will be the same value.

**Reply by mark2021blue** [reply]
Thanks for your answer!
Feel sorry! Maybe I didn't express or understand clearly, for a mnemonic, is the xpub of the default Derivation Path or the custom Derivation Path the same? If it is different, I want to calculate the child extended public key through TWHDWalletGetExtendedPublicKey and TWAccountExtendedPublicKey. I don't know what the problem is?
Thanks!

**Reply by mark2021blue** [reply]
The following interface parameters cannot specify custom Derivation Path
TWString *_Nonnull TWHDWalletGetExtendedPublicKey(struct TWHDWallet *_Nonnull wallet, enum TWPurpose purpose, enum TWCoinType coin, enum TWHDVersion version);

**Comment by mark2021blue** [comment]
The purpose of sharing the parent extended public key is so that child public keys (and thus addresses) can be derived from it，The default address is obtained by the following methods, and the derived address of the customDerivationPath2 cannot be obtained

TWPublicKey* pk1 = TWHDWalletGetPublicKeyFromExtended(TWAccountExtendedPublicKey(accountT1), TWCoinType::TWCoinTypeDogecoin, customDerivationPath2);
TWAnyAddress* pk1addr = TWAnyAddressCreateWithPublicKey(pk1, TWCoinType::TWCoinTypeDogecoin);
cout << "+++++++++++ pk1addr:" << TWStringUTF8Bytes(TWAnyAddressDescription(pk1addr)) << endl;

**Comment by mark2021blue** [comment]
The extendedPublicKey parameter in the following interface, I understand it is the master/parent extended public key, I want to obtain the child extended public key of the corresponding address of the derivationPath through TWAccountCreate, TWAccountExtendedPublicKey, etc.

struct TWAccount *_Nonnull TWAccountCreate(TWString *_Nonnull address, enum TWCoinType coin, TWString *_Nonnull derivationPath, TWString *_Nonnull extendedPublicKey);

**Reply by optout21** [reply]
The `Account` structure is only a grouping of public keys, derivation path, etc., but it does not perform any derivation.
I have not found a way to obtain extended HD private key for a custom derivation path.

**Reply by mark2021blue** [reply]
Wow, you are finally online, this series of problems has been bothering me for two days, and I need your help! ! !

I want to get the obtain extended HD public (not private key) key for a custom derivation path.

The purpose of sharing the parent extended public key is so that child public keys (and thus addresses) can be derived from it. How is this requirement achieved?
Thanks! ! !

**Reply by mark2021blue** [reply]
![2022-03-17 5 05 59](https://user-images.githubusercontent.com/83812275/158775410-3dc2732e-485b-459d-8429-4dfd9e432026.png)

**Reply by optout21** [reply]
As I written above, I have not found a way to obtain extended HD private key for a custom derivation path.
I have raised issue #2084 to add support for this.

**Reply by mark2021blue** [reply]
Thank you very much! ! !

**Reply by optout21** [reply]
Note: PR has been merged   https://github.com/trustwallet/wallet-core/pull/2085

**Reply by mark2021blue** [reply]
great news!
@catenocrypt @hewigovens  Thank you very much! ! !
